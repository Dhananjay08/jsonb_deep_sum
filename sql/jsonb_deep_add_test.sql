CREATE EXTENSION jsonb_deep_sum;
SELECT jsonb_deep_add('{"a": 2, "b": 1}', '{"a": 5}');
SELECT jsonb_deep_add('{"a": 2, "b": 3}', '{"a": 5, "b": 4}');
-- Objects with keys in different order
SELECT jsonb_deep_add('{"b": 2, "a": 3}', '{"a": 5, "b": 4}');
-- LHS key greater than RHS
SELECT jsonb_deep_add('{"b": 1}', '{"a": 2}');
-- RHS key greater than LHS
SELECT jsonb_deep_add('{"a": 2}', '{"b": 1}');
-- empty object
SELECT jsonb_deep_add('{}', '{"a": 2}');
-- not an object
SELECT jsonb_deep_add('1', '{"a": 1}');
-- NULL object
SELECT jsonb_deep_add(NULL, '{"a": 1}');
SELECT jsonb_deep_add('{"a": "bb"}', '{"b": 1}');
-- Nested and independent
SELECT jsonb_deep_add('{"a": {"a": 1}}', '{"b": 1}');
-- Deeply nested and independent
SELECT jsonb_deep_add('{"a": {"a": {"a": 1}}}', '{"b": 1}');
-- Nested reverse order
SELECT jsonb_deep_add('{"b": 1}', '{"a": {"a": 1}}');
-- Nested sum
SELECT jsonb_deep_add('{"a": {"a": 1}}', '{"a": {"a": 2}}');
-- Nested sum and additional property
SELECT jsonb_deep_add('{"a": {"a": 1}, "b": 1}', '{"a": {"a": 2}}');
-- Reversed nested sum and additional prooperty
SELECT jsonb_deep_add('{"a": {"a": 1}}', '{"a": {"a": 2}, "b": 1}');
-- Deeply nested sum
SELECT jsonb_deep_add('{"a": {"a": {"a": 4}}}', '{"a": {"a": {"a": 3}}}');
SELECT jsonb_deep_add('{"a": true}', '{"b": 1}');
SELECT jsonb_deep_add('{"a": [1, 2, 3,4]}', '{"b": 1}');
-- Incompatible objects
SELECT jsonb_deep_add('{"a": {"a": 1}}', '{"a": 1}');
SELECT jsonb_deep_add('{"a": 1}', '{"a": {"a": 1}}');


