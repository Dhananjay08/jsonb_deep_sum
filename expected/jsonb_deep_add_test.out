CREATE EXTENSION jsonb_deep_sum;
SELECT jsonb_deep_add('{"a": 2, "b": 1}', '{"a": 5}');
  jsonb_deep_add  
------------------
 {"a": 7, "b": 1}
(1 row)

SELECT jsonb_deep_add('{"a": 2, "b": 3}', '{"a": 5, "b": 4}');
  jsonb_deep_add  
------------------
 {"a": 7, "b": 7}
(1 row)

-- Objects with keys in different order
SELECT jsonb_deep_add('{"b": 2, "a": 3}', '{"a": 5, "b": 4}');
  jsonb_deep_add  
------------------
 {"a": 8, "b": 6}
(1 row)

-- LHS key greater than RHS
SELECT jsonb_deep_add('{"b": 1}', '{"a": 2}');
  jsonb_deep_add  
------------------
 {"a": 2, "b": 1}
(1 row)

-- RHS key greater than LHS
SELECT jsonb_deep_add('{"a": 2}', '{"b": 1}');
  jsonb_deep_add  
------------------
 {"a": 2, "b": 1}
(1 row)

-- empty object
SELECT jsonb_deep_add('{}', '{"a": 2}');
 jsonb_deep_add 
----------------
 {"a": 2}
(1 row)

-- not an object
SELECT jsonb_deep_add('1', '{"a": 1}');
ERROR:  Can only sum objects
-- NULL object
SELECT jsonb_deep_add(NULL, '{"a": 1}');
 jsonb_deep_add 
----------------
 
(1 row)

SELECT jsonb_deep_add('{"a": "bb"}', '{"b": 1}');
ERROR:  Only numeric values allowed
-- Nested and independent
SELECT jsonb_deep_add('{"a": {"a": 1}}', '{"b": 1}');
     jsonb_deep_add      
-------------------------
 {"a": {"a": 1}, "b": 1}
(1 row)

-- Deeply nested and independent
SELECT jsonb_deep_add('{"a": {"a": {"a": 1}}}', '{"b": 1}');
         jsonb_deep_add         
--------------------------------
 {"a": {"a": {"a": 1}}, "b": 1}
(1 row)

-- Nested reverse order
SELECT jsonb_deep_add('{"b": 1}', '{"a": {"a": 1}}');
     jsonb_deep_add      
-------------------------
 {"a": {"a": 1}, "b": 1}
(1 row)

-- Nested sum
SELECT jsonb_deep_add('{"a": {"a": 1}}', '{"a": {"a": 2}}');
 jsonb_deep_add  
-----------------
 {"a": {"a": 3}}
(1 row)

-- Nested sum and additional property
SELECT jsonb_deep_add('{"a": {"a": 1}, "b": 1}', '{"a": {"a": 2}}');
     jsonb_deep_add      
-------------------------
 {"a": {"a": 3}, "b": 1}
(1 row)

-- Reversed nested sum and additional prooperty
SELECT jsonb_deep_add('{"a": {"a": 1}}', '{"a": {"a": 2}, "b": 1}');
     jsonb_deep_add      
-------------------------
 {"a": {"a": 3}, "b": 1}
(1 row)

-- Deeply nested sum
SELECT jsonb_deep_add('{"a": {"a": {"a": 4}}}', '{"a": {"a": {"a": 3}}}');
     jsonb_deep_add     
------------------------
 {"a": {"a": {"a": 7}}}
(1 row)

SELECT jsonb_deep_add('{"a": true}', '{"b": 1}');
ERROR:  Only numeric values allowed
SELECT jsonb_deep_add('{"a": [1, 2, 3,4]}', '{"b": 1}');
       jsonb_deep_add        
-----------------------------
 {"a": [1, 2, 3, 4], "b": 1}
(1 row)

-- Incompatible objects
SELECT jsonb_deep_add('{"a": {"a": 1}}', '{"a": 1}');
ERROR:  Incompatible types
SELECT jsonb_deep_add('{"a": 1}', '{"a": {"a": 1}}');
ERROR:  Incompatible types
